cache/cache.go

```diff
@@ -13,3 +13,284 @@
+	select {
+	case <-c.ready:
+	case <-ctx.Done():
+		emptyWatchChan := make(chan clientv3.WatchResponse)
+		close(emptyWatchChan)
+		return emptyWatchChan
+	}
+
+	if !strings.HasPrefix(key, c.prefix) {
+		emptyWatchChan := make(chan clientv3.WatchResponse)
+		close(emptyWatchChan)
+		return emptyWatchChan
+	}
+
+	op := clientv3.OpGet(key, opts...)
+	requestedRev := op.Rev()
+
+	responseChan := make(chan clientv3.WatchResponse, 1)
+
+	// build the internal Watch event stream (for historic reply + live)
+	stream, err := c.newWatchEventStream(ctx, key, requestedRev)
+	if err != nil {
+		if errors.Is(err, ErrCompacted) {
+			var compactRev int64
+			if oldestEvent := c.history.PeekOldest(); oldestEvent != nil {
+				compactRev = oldestEvent.Kv.ModRevision
+			}
+			responseChan <- clientv3.WatchResponse{CompactRevision: compactRev}
+		} else {
+			responseChan <- clientv3.WatchResponse{Canceled: true}
+		}
+		close(responseChan)
+		return responseChan
+	}
+
+	go func() {
+		defer close(responseChan)
+		for {
+			select {
+			case event, ok := <-stream:
+				if !ok {
+					return
+				}
+				responseChan <- clientv3.WatchResponse{Events: []*clientv3.Event{event}}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return responseChan
+}
+
+// Ready reports whether the cache has finished its initial load.
+func (c *Cache) Ready() bool {
+	select {
+	case <-c.ready:
+		return true
+	default:
+		return false
+	}
+}
+
+// WaitReady blocks until the cache is ready or the ctx is cancelled.
+func (c *Cache) WaitReady(ctx context.Context) error {
+	select {
+	case <-c.ready:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Close cancels the private context and blocks until all goroutines return.
+func (c *Cache) Close() {
+	c.stop()
+	c.waitGroup.Wait()
+}
+
+// newWatchEventStream builds a single internal Watch event stream for
+// both historical replay (>= startRev) and live updates filtered by key.
+func (c *Cache) newWatchEventStream(ctx context.Context, key string, startRev int64) (WatchEventStream, error) {
+	var pred KeyPredicate
+	if key == "\x00" { // special case for "\x00" which means “entire key-space”
+		pred = nil // accept every event
+	} else {
+		pred = func(k []byte) bool { return strings.HasPrefix(string(k), key) }
+	}
+
+	// TODO: re-enable arbitrary startRev support once we guarantee gap-free replay.
+	if startRev != 0 {
+		return nil, ErrUnsupportedStartRev
+	}
+
+	var nextRev int64
+	if latestEvent := c.history.PeekLatest(); latestEvent != nil {
+		nextRev = latestEvent.Kv.ModRevision + 1
+	}
+
+	w := newWatcher(c.cfg.PerWatcherBufferSize, pred)
+	c.demux.add(w, nextRev)
+
+	go func() {
+		<-ctx.Done()
+		c.demux.remove(w)
+	}()
+
+	return w.eventQueue, nil
+}
+
+func serveWatchEvents(ctx context.Context, watchCtx *watchCtx, bufSize int) {
+	enqueueCh := make(chan *clientv3.Event, bufSize)
+
+	go func() {
+		for event := range enqueueCh {
+			watchCtx.broadcast(event)
+		}
+	}()
+
+	// upstream watch/retry loop
```

Comment:
Not needed comment `upstream watch/retry loop` for function named `watchRetryLoop`. If you think `upstream` is important here, just call function `upstreamWatchRetryLoop`, but don't write non informative comments.

Good practice: Use comments to express WHY? and not WHAT?. This works much better because you it's hard to name/structure code to express WHY?, but pretty easy to for WHAT?